aten/src/ATen/native/IndexingUtils.h

```diff
@@ -39,9 +40,15 @@ static void invalid_mask(const Tensor & self, int64_t idx, const Tensor & mask,
         }
         // Replace with nonzeros
         auto nonzero = index.nonzero();
+        if (ensure_same_device && nonzero.device() != self.device()) {
+          bool non_blocking = nonzero.is_cpu() && self.device().is_cuda();
```

Comment:
this is not enough to make a copy non-blocking, you also need to copy nonzero to pinned memory

Reply:
Thanks for the fast review. I verified in the pytorch profile that `cudaStreamSynchronize` is not called even with this code and `test_sync_warning` also checks that.

How would I create a `nonzero` in pinned memory? Can I just create an zero sized empty tensor in pinned memory and use that as the output parameter or is that problematic since I don't know the tensorsize beforehand?

I'm just asking because I believe simply calling [`nonzero.pin_memory()` is not recommended ](https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html#what-you-will-learn) or is this different here?

Reply:
You could allocate zero-element pinned tensor and send it as `out` arg to `nonzero`